### PR TITLE
Write total_operating_time and cycles_numbers to EEPROM every 5 min #7

### PR DESCRIPTION
**Pull request by jonadem**
To test it, it is easier to change WRITE_EEPROM_PERIOD_MS from "3000 * 1000L" to "5 * 1000L".

incrementCyclesNumber() need to be called from another task (and the current call need to be removed [there is a FIXME]).

### CODE REVIEW

**Comment by anpar** [comment]
Looks good! Will be tested today.

File: src/core/main_task.c [hunk 1]
```diff
@@ -29,6 +36,9 @@ uint8_t extra_param;
 static void process_alarm(uint32_t notification);
 static void process_critical_failure(uint32_t notification);
 
+#define WRITE_EEPROM_PERIOD_MS 3000 * 1000L // Try 5 * 1000L for Testing purpose
+#define EEPROM_INIT_MAGIC_NUMBER 1
```
Comment (cassiersg) [inline src/core/main_task.c:40]: Use a random number as magi number (othewise there is a risk of finding it correct by chance).
```suggestion
#define EEPROM_INIT_MAGIC_NUMBER 0xc417b8c6
```

File: src/core/main_task.c [hunk 2]
```diff
@@ -63,6 +73,31 @@ void MainTask(void *pvParameters)
...
+        debug_print("[MAIN] EEPROM, tot: %u\r\n", total_operating_time);
+        debug_print("[MAIN] EEPROM, cn: %u\r\n", cycles_number);
+#endif
+    }
+    else {
+        //First boot ever
+        total_operating_time = 0;
+        cycles_number = 0;
+        taskENTER_CRITICAL();
+        eeprom_write_byte(&eeprom_init_flag_addr, EEPROM_INIT_MAGIC_NUMBER);
```
Comment (cassiersg) [inline src/core/main_task.c:97]: Shouldn't we first initialize to zero the counters before writing the magic number & make it dword ?
```suggestion
eeprom_write_dword(operating_time_addr, 0);
eeprom_write_dword(cycles_number_addr, 0);
eeprom_write_dword(eeprom_init_flag_addr, EEPROM_INIT_MAGIC_NUMBER);
` ``

File: src/core/main_task.c [hunk 3]
```diff
@@ -372,8 +407,19 @@ void MainTask(void *pvParameters)
...
+            total_operating_time += WRITE_EEPROM_PERIOD_MS/1000L;
+            incrementCyclesNumber(); // FIXME Remove this (should be called by another task)
+#if DEBUG_MAIN
+            debug_print("[MAIN] EEPROM, tot: %u\r\n", total_operating_time);
+            debug_print("[MAIN] EEPROM, cn: %u\r\n", cycles_number);
+#endif
+            taskENTER_CRITICAL();
+            eeprom_write_dword (&operating_time_addr, total_operating_time);
+            eeprom_write_dword (&cycles_number_addr, cycles_number);
+            taskEXIT_CRITICAL();
```
Comment (cassiersg) [inline src/core/main_task.c:421]: Could this be split into two distinct critical sections ?
```suggestion
            taskENTER_CRITICAL();
            eeprom_write_dword (&operating_time_addr, total_operating_time);
            taskEXIT_CRITICAL();
            taskENTER_CRITICAL();
            eeprom_write_dword (&cycles_number_addr, cycles_number);
            taskEXIT_CRITICAL();
```

**Comment by anpar** [comment]
Seems to work for reading/writing and EEPROM but breaks the LCD at some point for a yet unknown reason.

File: src/core/main_task.c [hunk 4]
```diff
@@ -63,6 +73,31 @@ void MainTask(void *pvParameters)
 {
     TickType_t xLastWakeTime = xTaskGetTickCount();
 
+    write_time = xTaskGetTickCount();
+
+    uint8_t eeprom_init_flag_addr = 0;
+    uint32_t operating_time_addr = 1;
+    uint32_t cycles_number_addr = 5;
```
Comment (cassiersg) [inline src/core/main_task.c:80]: Addresses should be pointers:
```suggestion
    uint32_t *eeprom_init_flag_addr = (uint32_t *) 0;
    uint32_t *operating_time_addr = (uint32_t *) 4;
    uint32_t *cycles_number_addr = (uint32_t *) 8;
```

**Comment by cassiersg** [comment]
> Seems to work for reading/writing EEPROM but breaks the LCD at some point for a yet unknown reason.
> 
> 

Probably due to disabling interrupts for too long. (Maybe try without the taskEnterCritical / taskExitCritical ?)

**Comment by anpar** [comment]
> > Seems to work for reading/writing EEPROM but breaks the LCD at some point for a yet unknown reason.
> > 
> 
> Probably due to disabling interrupts for too long. (Maybe try without the taskEnterCritical / taskExitCritical ?)

I will. But for now I'm struggling to reproduce the bug while I haven't changed anything. Weird.

**Pull request merged**